File: simulation/weather.py
"""Seasonal deterministic weather generation and plot updates."""

from simulation import crop_growth, derived

SEASONS = ("spring", "summer", "autumn", "winter")

# Optional compiled accelerator for the per-plot loop below (~20% of batch
# runtime). Absent unless `python3 setup.py build_ext --inplace` has been
# run, in which case the pure-Python loop -- which stays the reference
# implementation -- is used instead. The layout check makes a stale .so
# built against a different CropProfile.flat fall back rather than misread
# it. tests/test_fastplot_equivalence.py asserts the two agree bit-for-bit.
try:
    from simulation import _fastplot
except ImportError:
    _fastplot = None
else:
    if getattr(_fastplot, "PROFILE_LAYOUT", None) != derived.PROFILE_LAYOUT:
        _fastplot = None


def season_for_day(day: int, season_length: int = 15) -> str:
    return SEASONS[(day // season_length) % 4]


def generate_weather(day: int, config: dict, rng) -> dict:
    # Season parameters are fixed config, resolved once per config object
    # rather than re-read (with defaults) on every simulated day. The rng call
    # order below is load-bearing and unchanged: temperature draw, then the
    # rain check, then the rainfall draw only if it rained.
    params = derived.weather_params(config)
    season = SEASONS[(day // params.season_length) % 4]
    temp_low, temp_high, rain_chance, rain_low, rain_high, base_evaporation = params.by_season[
        season
    ]
    temperature = rng.uniform(temp_low, temp_high)
    rainfall = rng.uniform(rain_low, rain_high) if rng.chance(rain_chance) else 0.0
    evaporation = base_evaporation + max(0.0, temperature - 25) * 0.005
    return {
        "season": season,
        "temperature": round(temperature, 2),
        "rainfall": round(rainfall, 3),
        "evaporation": round(evaporation, 3),
    }


def apply_weather(
    player,
    crops_by_id: dict,
    weather: dict,
    growth_module,
    crop_profiles=None,
    plot_regen=None,
    dynamics=None,
    crop_profiles_flat=None,
) -> None:
    # Weather values are the same for every plot, so they are read once here
    # rather than per plot. `crop_profiles` maps crop_id to its cached static
    # growth inputs; the engine passes the one it already holds, and omitting
    # it just falls back to an equivalent per-crop lookup.
    rainfall = weather.get("rainfall", 0.0)
    # Evaporation is a property of the day's weather, not of whether a plot
    # happens to be occupied, so it is read once here and applied to every
    # plot. Growing plots have it applied by growth_module.update_crop_stress
    # (after that function reads today's moisture for water stress); fallow
    # plots have it applied in the `planted is None` branch below. Previously
    # only the former happened, so a fallow plot took on rainfall and never
    # gave any back -- it saturated at 1.0 and handed the next crop planted
    # there several stress-free days that no design doc ever granted it.
    evaporation = weather.get("evaporation", 0.08)
    day = player.day
    dynamics = dynamics if dynamics is not None else derived.DEFAULT_DYNAMICS
    # Resolved once per day, not per plot -- `plot_regen` is already a
    # cached, per-world dict (see derived.WorldLookups.plot_regen), and
    # defaults to no regen so a caller that omits it keeps today's behaviour.
    # Covers nitrogen/phosphorus/potassium plus soil_health/pest_pressure/
    # disease_pressure -- all four of these only otherwise recover when a
    # plot sits completely fallow (see the `planted is None` branch below),
    # which no shipped agent strategy ever deliberately does, so under
    # continuous farming every one of them used to march to its worst value
    # and stay there. This makes that recovery a small, constant trickle
    # regardless of occupancy; the fallow-only bonuses below still stack on
    # top, so deliberate rest/rotation keeps paying off faster.
    regen_moisture = plot_regen.get("moisture", 0.0) if plot_regen else 0.0
    regen_n = plot_regen.get("nitrogen", 0.0) if plot_regen else 0.0
    regen_p = plot_regen.get("phosphorus", 0.0) if plot_regen else 0.0
    regen_k = plot_regen.get("potassium", 0.0) if plot_regen else 0.0
    regen_soil_health = plot_regen.get("soil_health", 0.0) if plot_regen else 0.0
    regen_pest = plot_regen.get("pest_pressure", 0.0) if plot_regen else 0.0
    regen_disease = plot_regen.get("disease_pressure", 0.0) if plot_regen else 0.0
    regenerates_nutrients = regen_n or regen_p or regen_k

    # Compiled fast path, taken only when it is a like-for-like substitution:
    # the caller must be using the real crop_growth module (a few tests pass
    # their own) and must already hold the flattened per-crop profiles the
    # kernel indexes. Anything else falls through to the loop below.
    if _fastplot is not None and growth_module is crop_growth and crop_profiles_flat is not None:
        _fastplot.apply_day(
            player.plots,
            day,
            rainfall,
            evaporation,
            weather.get("temperature", 20),
            (
                regen_moisture,
                regen_n,
                regen_p,
                regen_k,
                regen_soil_health,
                regen_pest,
                regen_disease,
            ),
            (
                dynamics.fallow_pest_decay,
                dynamics.fallow_disease_decay,
                dynamics.fallow_soil_health_regen,
                dynamics.max_disease_pressure,
                dynamics.disease_growth_per_rainfall,
                dynamics.max_pest_pressure,
                dynamics.pest_growth_per_day,
            ),
            crop_profiles_flat,
        )
        return

    for plot in player.plots:
        # Config-accepted (soil.regen_per_day.moisture) alongside the other
        # SOIL_LEVELS regen fields, folded into the same rainfall addition
        # since both are a same-day, once-capped increase to moisture.
        plot.moisture = min(1.0, plot.moisture + rainfall + regen_moisture)
        if regenerates_nutrients:
            # A crop's own demand (see crop_growth.update_crop_stress) still
            # outpaces this for nutrient-hungry crops under sustained
            # monocropping; it only caps runaway depletion, it doesn't erase
            # the cost of farming.
            if regen_n:
                plot.nitrogen = min(1.0, plot.nitrogen + regen_n)
            if regen_p:
                plot.phosphorus = min(1.0, plot.phosphorus + regen_p)
            if regen_k:
                plot.potassium = min(1.0, plot.potassium + regen_k)
        if regen_soil_health:
            plot.soil_health = min(1.0, plot.soil_health + regen_soil_health)
        if regen_pest:
            plot.pest_pressure = max(0.0, plot.pest_pressure - regen_pest)
        if regen_disease:
            plot.disease_pressure = max(0.0, plot.disease_pressure - regen_disease)
        planted = plot.crop
        if planted is None:
            plot.moisture = max(0.0, min(1.0, plot.moisture - evaporation))
            plot.pest_pressure = max(0.0, plot.pest_pressure * dynamics.fallow_pest_decay)
            plot.disease_pressure = max(0.0, plot.disease_pressure * dynamics.fallow_disease_decay)
            plot.soil_health = min(1.0, plot.soil_health + dynamics.fallow_soil_health_regen)
            continue
        crop_id = planted.crop_id
        crop = crops_by_id[crop_id]
        growth_module.update_crop_stress(
            planted,
            plot,
            crop,
            weather,
            crop_profiles[crop_id] if crop_profiles is not None else None,
        )
        interval = crop.get("water_interval_days", 3)
        overdue = day - planted.last_watered_day - interval
        planted.neglect_days = max(0, overdue)
        plot.disease_pressure = min(
            dynamics.max_disease_pressure,
            plot.disease_pressure + rainfall * dynamics.disease_growth_per_rainfall,
        )
        plot.pest_pressure = min(
            dynamics.max_pest_pressure, plot.pest_pressure + dynamics.pest_growth_per_day
        )

File: farm-c/src/weather.c
#include "weather.h"

#include "crop_growth.h"
#include "pyfloat.h"

/* simulation/weather.py:22-23. */
Season weather_season_for_day(int day, int season_length_days) {
    /* Python's `//` and `%` on non-negative operands agree with C's `/` and
     * `%`; day and season_length_days are both always non-negative in this
     * port's callers (day counts up from 0, season_length_days is a
     * positive config value), so no floor-division helper is needed here
     * unlike contracts.c/profit_optimizer.c's genuinely-signed cases. */
    return (Season)((day / season_length_days) % SEASON_COUNT);
}

/* simulation/weather.py:26-44. */
WeatherDay weather_generate(int day, const ResolvedConfig *config, FarmRng *rng) {
    const WeatherParams *params = &config->weather;
    Season season = weather_season_for_day(day, params->season_length_days);
    const SeasonWeather *by_season = &params->by_season[season];

    /* Draw order fixed: temperature, then the rain-chance roll, then
     * (conditionally) the rainfall draw -- see weather.h's header comment. */
    double temperature = rng_uniform_event(rng, by_season->temperature_low, by_season->temperature_high);
    double rainfall = rng_chance(rng, by_season->rain_chance)
                           ? rng_uniform_event(rng, by_season->rainfall_low, by_season->rainfall_high)
                           : 0.0;
    /* Computed from the raw (unrounded) temperature draw, before rounding
     * is applied to any of the three output fields below -- matching
     * weather.py's `evaporation = base_evaporation + max(0.0, temperature -
     * 25) * 0.005` line, which runs before the returned dict rounds
     * anything. */
    double evaporation = by_season->evaporation + py_max(0.0, temperature - 25) * 0.005;

    WeatherDay result;
    result.season = season;
    result.temperature = py_round_ndigits(temperature, 2);
    result.rainfall = py_round_ndigits(rainfall, 3);
    result.evaporation = py_round_ndigits(evaporation, 3);
    return result;
}

/* simulation/weather.py:47-174 (plain loop; see weather.h's header comment
 * for why this ports that path rather than _fastplotmodule.c's). */
void weather_apply(FarmState *state, const WeatherDay *weather) {
    const ResolvedConfig *config = state->config;
    const SoilDynamics *dynamics = &config->soil_dynamics;
    const PlotRegen *regen = &config->plot_regen;
    double rainfall = weather->rainfall;
    double evaporation = weather->evaporation;
    int day = state->day;

    /* Python: `regen_n or regen_p or regen_k` -- a truthiness gate only,
     * evaluated once per day rather than once per plot per nutrient. */
    bool regenerates_nutrients =
        (regen->nitrogen != 0.0) || (regen->phosphorus != 0.0) || (regen->potassium != 0.0);

    for (size_t i = 0; i < state->plots.count; i++) {
        /* weather.py:130: `plot.moisture = min(1.0, plot.moisture +
         * rainfall + regen_moisture)` -- left-associative, matching C's
         * default grouping for `a + b + c`. Direct column access here (not
         * a gathered PlotState local) is the actual cache-locality win of
         * the SoA layout for this, the hottest per-day loop. */
        state->plots.moisture[i] =
            py_min(1.0, state->plots.moisture[i] + rainfall + regen->moisture);
        if (regenerates_nutrients) {
            if (regen->nitrogen != 0.0) {
                state->plots.nitrogen[i] = py_min(1.0, state->plots.nitrogen[i] + regen->nitrogen);
            }
            if (regen->phosphorus != 0.0) {
                state->plots.phosphorus[i] =
                    py_min(1.0, state->plots.phosphorus[i] + regen->phosphorus);
            }
            if (regen->potassium != 0.0) {
                state->plots.potassium[i] =
                    py_min(1.0, state->plots.potassium[i] + regen->potassium);
            }
        }
        if (regen->soil_health != 0.0) {
            state->plots.soil_health[i] = py_min(1.0, state->plots.soil_health[i] + regen->soil_health);
        }
        if (regen->pest_pressure != 0.0) {
            state->plots.pest_pressure[i] =
                py_max(0.0, state->plots.pest_pressure[i] - regen->pest_pressure);
        }
        if (regen->disease_pressure != 0.0) {
            state->plots.disease_pressure[i] =
                py_max(0.0, state->plots.disease_pressure[i] - regen->disease_pressure);
        }

        if (state->plots.planted_index[i] < 0) {
            /* Fallow: these four fields are always written, unlike the
             * conditional regen writes above (matches weather.py:150-153
             * running unconditionally once a plot is confirmed fallow). */
            state->plots.moisture[i] = clamp01(state->plots.moisture[i] - evaporation);
            state->plots.pest_pressure[i] =
                py_max(0.0, state->plots.pest_pressure[i] * dynamics->fallow_pest_decay);
            state->plots.disease_pressure[i] =
                py_max(0.0, state->plots.disease_pressure[i] * dynamics->fallow_disease_decay);
            state->plots.soil_health[i] =
                py_min(1.0, state->plots.soil_health[i] + dynamics->fallow_soil_health_regen);
            continue;
        }

        size_t planted_index = (size_t)state->plots.planted_index[i];
        const CropDef *crop = config_find_crop(config, state->planted.crop_item_id[planted_index]);

        /* crop_growth_update_stress mutates both its PlantedCrop and
         * PlotState in place, and its signature stays plain pointers
         * (tests/test_physics.c exercises it standalone with no FarmState)
         * -- gather this one plot's and one crop's rows, call, scatter the
         * mutated rows back. */
        PlotState plot_row = plot_columns_get(&state->plots, i);
        PlantedCrop planted_row = planted_crop_columns_get(&state->planted, planted_index);
        crop_growth_update_stress(&planted_row, &plot_row, crop, weather->temperature, evaporation);
        plot_columns_set(&state->plots, i, plot_row);
        planted_crop_columns_set(&state->planted, planted_index, planted_row);

        int overdue =
            day - state->planted.last_watered_day[planted_index] - crop->water_interval_days;
        state->planted.neglect_days[planted_index] = overdue > 0 ? overdue : 0;

        state->plots.disease_pressure[i] = py_min(
            dynamics->max_disease_pressure,
            state->plots.disease_pressure[i] + rainfall * dynamics->disease_growth_per_rainfall);
        state->plots.pest_pressure[i] = py_min(
            dynamics->max_pest_pressure, state->plots.pest_pressure[i] + dynamics->pest_growth_per_day);
    }
}
